RFC: Add more support for fallible allocations in Vec (#3271)
rust-lang/rfcs: text/0000-vec-fallible-allocation.md

- Feature Name: `vec_fallible_allocation`
- Start Date: 2022-20-04
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

`Vec` has many methods that may allocate memory (to expand the underlying storage, or shrink it down). Currently each of these method rely on "infallible"
allocation, that is any failure to allocate will call the global OOM handler, which will (typically) panic. Even if the global OOM handler does not panic, the
return type of these method don't provide a way to indicate the failure.

Currently `Vec` does have a `try_reserve` method that uses "fallible" allocation: if `try_reserve` attempts to allocate, and that allocation fails, then the
return value of `try_reserve` will indicate that there was a failure, and the `Vec` is left unchanged (i.e., in a valid, usable state). We propose adding

> more of these `try_` methods to `Vec`, specifically for any method that can allocate. However, unlike most RFCs, we are not suggesting that this proposal is

Reviewer: ```suggestion
more of these `try_` methods to `Vec`, specifically for any method that can allocate.

Unlike most RFCs, we are not suggesting that this proposal is
```
I would make this a new paragraph

> the best among many alternatives (in fact we know that adding more `try_` methods is undesirable in the long term), instead we are suggesting this as a way

Reviewer 2: What is the rationale behind the "we know"? i.e. why do we know?

Reviewer 2: Thanks!

Yeah, the other section explains the drawbacks, but having drawbacks does not necessarily mean the methods need to be introduced with the expectation of getting removed. Maybe it would be good to explain why we want to introduce this as a temporary solution.

Author: I'd rather point folks at the drawbacks and motivations section than increase the summary section any further: it's already larger than the suggested size and I'd only be repeating what the other two sections say.

> the best among many alternatives (in fact we know that adding more `try_` methods is undesirable in the long term), instead we are suggesting this as a way
> forward to unblock important work (see the "Motivations" section below) while we explore other alternatives.

Reviewer: ```suggestion
the best among all current and futures alternatives (in fact we know that adding more `try_` methods is undesirable in the long term). Instead we are suggesting this as the best alternative *not requiring language changes*, and thus as a way forward to unblock important work (see the "Motivations" section below) while we explore other alternatives.
```
something like that. Maybe spell out "no plans of stabilizing this in its current form" too?

    pub fn try_into_boxed_slice(self) -> Result<Box<[T], A>, TryReserveError>;
    pub fn try_push(&mut self, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize(&mut self, new_len: usize, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize_with<F>(&mut self, new_len: usize, f: F) -> Result<(), TryReserveError> where F: FnMut() -> T;
    pub fn try_shrink_to(&mut self, min_capacity: usize) -> Result<(), TryReserveError>;
    pub fn try_shrink_to_fit(&mut self) -> Result<(), TryReserveError>;
    pub fn try_split_off(&mut self, at: usize) -> Result<Self, TryReserveError> where A: Clone; // NOTE: still panics if given an invalid index
    pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
}

#[doc(hidden)]
pub fn try_from_elem<T: Clone>(elem: T, n: usize) -> Result<Vec<T>, TryReserveError>;

#[doc(hidden)]
pub fn try_from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Result<Vec<T, A>, TryReserveError>;

```

# Drawbacks
[drawbacks]: #drawbacks

Bifurcating the API surface like this is undesirable: it adds yet another decision point for developers using a fundamental type, and requires the developer to

> stop and understand what could possible fail in the `try_` version. Additionally, this bifurcation can't stop at just `Vec`: What about the rest of the

Reviewer 2: ```suggestion
stop and understand what could possibly fail in the `try_` version. Additionally, this bifurcation can't stop at just `Vec`: What about the rest of the
```

Disadvantages:

- Very easy to forget to call the side-channel method, and not doing so is highly likely to lead to hard-to-diagnose bugs (e.g., code that thinks it has pushed
  an item but actually hasn't).
  - Static analysis could be added to assist with detecting this.
  - Existing code is unaware of the side-channel, and so will never call it.
  - Subsequent calls to the `Vec` could panic if the last allocation failed, but this would be a performance hit and might not help diagnosability (the
    developer would see a crash at the subsequent call, not the call that failed to allocate).
- Requires that developers add a lot of boilerplate code, slightly less if the side-channel returns a `Result` that the developer can use `?` on.
- It is not obvious if an API allocates or not, so even if a developer is aware of the side-channel method they may not think it needs to be called, or may call
  it too often.

### Add a trait for the "fallible allocation" functions (to effectively hide them)

We could create a new trait specifically for the `try_` methods in `Vec` - this will effectively hide them unless a developer is looking through the list of

> trait implementations for `Vec` or adds a `use` declaration for the trait.

Reviewer 2: If the idea here is to hide them, then I think a documentation-level feature would be better (and could be used for other purposes), e.g. https://github.com/rust-lang/rust/pull/95051#issuecomment-1128639131.

Of course, those may not exist yet, but this RFC could be a good rationale for such kind of feature.

trait implementations for `Vec` or adds a `use` declaration for the trait.

Advantages:

- Although we are still adding new methods to `Vec`, they will be hidden from most developers and so avoids the cognitive burden that comes from them being
  directly on the type.
- The trait and the additional methods could be put in a standalone crate (it would require duplicate quite a bit of code and relying on `unsafe` functions
  like `set_len`).

Disadvantages:

- Still requires bifurcating traits, and the functions that call those traits.
- Unusual use for a trait, and would require additional documentation to explain why these methods have been implemented this way.
- Any future allocating methods will require a new trait to implement them, since adding functions to a trait is a breaking change.

### Use `#[doc(hidden)]` (or a new doc feature) to hide the methods

We could mark all of the new methods with `#[doc(hidden)]` to hide them in the documentation, or perhaps introduce a new doc attribute to place these methods
in their own section (collapsed by default?). Having documentation sections could be made even more useful by adding syntax to have documentation at the top of

> that section.

Reviewer 2: Note that this would not need to be an alternative as such, since it is an orthogonal feature that can be used to improve the one suggested in the RFC -- which is a big plus!

Instead, I would put it in "Future possibilities" and describe it as a potential way to reduce the drawbacks of the current solution proposed by the RFC.

Author: Ok, moved the discussion about doc sections into "Future possibilities"

below) while we explore other alternatives. We have no plans to stabilize this feature in its current form and fully expect that these methods will be removed
in the future.

# Motivation
[motivation]: #motivation

The motivation for this change is well documented by other RFCs, such as the proposed [fallible_allocation feature](https://github.com/rust-lang/rfcs/pull/3140)
and the accepted [fallible_collection_alloc feature](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md).

As a brief summary, there are environments where dynamic allocation is required and failing to allocate is both expected and able to be handled. For example,
in OS Kernel environments (such as [Linux](https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/)), some embedded
systems, high-reliability systems (such as databases) and multi-user services (where a single request may fail without the entire service halting, such as a
web server).

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> One often unconsidered failure in a software system is running out of memory (commonly referred to as "Out-of-memory" or OOM): if code attempts to dynamically
> allocate memory, and there is an insufficient amount of available memory on the system, how should this be handled? In most applications, the failure to
> allocate means that the code cannot make any progress, and so the appropriate response is to exit immediately indicating that there was a failure (i.e., to
> `panic`). Consider the Rust Compiler itself: if the compiler cannot allocate memory then it cannot fail that part of the compilation and attempt to continue
> the rest of the compilation - the only appropriate action is to fail the entire compilation and emit an error message. Since panicking on allocation failure is
> appropriate in most circumstances, this is the default within the Rust standard library. Confusingly, this approach is referred to as "infallible allocation":
> from the perspective of something calling a function it can assume that allocations cannot fail (otherwise the entire process will be terminated).
> 
> There are, however, environments where allocation failures are both expected and able to be handled. Consider a multi-user service such as a database or a web
> server: if a request comes in to the service that requires allocating more memory than is available on the system, then the service can respond with a failure
> for just that request, and it can continue to service other requests. This approach is referred to as "fallible allocation": each function that may allocate
> needs to indicate to its caller if an attempt to allocate has failed. Within the Rust standard library, one can identify "fallible allocation" functions by
> their names being prefixed with `try_` and their return types being an instance of `Result`.

Reviewer: Tried to break up into more paragraphs. I also like doing once sentence per line because it makes for nice diffs (after the switch!) Conversely, one can wrap at fixed length, have nice diffs, but not both.

```suggestion
One often unconsidered failure in a software system is running out of memory (commonly referred to as "Out-of memory" or OOM): 
if code attempts to dynamically allocate memory, and there is an insufficient amount of available memory on the system, how should this be handled?

In most applications, the failure to allocate means that the code cannot make any progress, and so an appropriate response is to exit immediately indicating that there was a failure (i.e., to `panic`).
Consider the Rust Compiler itself: if the compiler cannot allocate memory then it cannot fail that part of the compilation and attempt to continue the rest of the compilation --- the only appropriate action is to fail the entire compilation and emit an error message.

Panicking on allocation failureis the default within the Rust standard library for these reasons.
This approach is referred to as "infallible allocation", rather confusingly.
The name is from the perspective of something calling a function.
The caller can assume that allocations cannot fail, for if it indeed doesn't the assumption is valid, and if it in fact the does then the program panics and the caller will not live to observe the failure.

There are, however, environments where allocation failures are expected and also required to be handled more gracefully.
Consider a multi-user service such as a database or a web server:
if a request comes in to the service that requires allocating more memory than is available on the system, then the service can respond with a failure for just that request, and it can continue to service other requests.
This approach is referred to as "fallible allocation".
Each function that may allocate needs to indicate to its caller if an attempt to allocate has failed, and the caller needs to explicitly handle that possibility.

Within the Rust standard library, one can identify "fallible allocation" functions by their names being prefixed with `try_` and their return types being an instance of `Result`.
```

Author: I've split up the paragraphs and taken some of your wording (thanks!).
I wasn't sure if there was a standard for formatting markdown files, but splitting per-sentence to make the diffs nicer makes sense.

Reviewer: Looks good! Feel free to mark this thread resolved.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> Currently any function in `alloc` that may allocate is marked with `#[cfg(not(no_global_oom_handling))]` (see <https://github.com/rust-lang/rust/pull/84266>),

Reviewer: ```suggestion
Currently any function in `alloc` that may call the global oom handler (which panics) rather than let its caller handle all allocation failures is marked with `#[cfg(not(no_global_oom_handling))]` (see <https://github.com/rust-lang/rust/pull/84266>),
```

    pub fn try_insert(&mut self, index: usize, element: T) -> Result<(), TryReserveError>; // NOTE: still panics if given an invalid index
    pub fn try_into_boxed_slice(self) -> Result<Box<[T], A>, TryReserveError>;
    pub fn try_push(&mut self, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize(&mut self, new_len: usize, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize_with<F>(&mut self, new_len: usize, f: F) -> Result<(), TryReserveError> where F: FnMut() -> T;
    pub fn try_shrink_to(&mut self, min_capacity: usize) -> Result<(), TryReserveError>;
    pub fn try_shrink_to_fit(&mut self) -> Result<(), TryReserveError>;
    pub fn try_split_off(&mut self, at: usize) -> Result<Self, TryReserveError> where A: Clone; // NOTE: still panics if given an invalid index
    pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
}

#[doc(hidden)]
pub fn try_from_elem<T: Clone>(elem: T, n: usize) -> Result<Vec<T>, TryReserveError>;

#[doc(hidden)]
pub fn try_from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Result<Vec<T, A>, TryReserveError>;

```

> 

Reviewer: ```suggestion
## Stability / lifecycle

The `try_` methods are added without any presumption that they will eventually be stabilized in their current form, and this will be documented.
This is a compromise between those that need this functionality now, and those that are wary of the `try_*` ad-hoc namespacing being "grandfathered in" before a future superior solution is implemented.

```

# Drawbacks
[drawbacks]: #drawbacks

Bifurcating the API surface like this is undesirable: it adds yet another decision point for developers using a fundamental type, and requires the developer to
stop and understand what could possibly fail in the `try_` version. Additionally, this bifurcation can't stop at just `Vec`: What about the rest of the
collection types? What about the traits that `Vec` implements (e.g., `IntoIter`)? What about other types that implement those traits? What about functions that
use those traits? This bifurcation becomes viral both within the standard library and to other crates as well.

Implementing both functions using a generic implementation method also complicates the code within `Vec`, making it harder to understand and maintain. While
there is some complex code within `RawVec`, the code in `Vec` itself tends to be fairly straightforward and within reason for an inexperienced developer to
understand.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Rationale

> Adding methods with the `try_` prefix follows the existing pattern within the standard library (as used by `Box::try_new` and `Vec::try_reserve`) and doesn't
> require any changes to the language or the compiler. Although this design does bifurcate the functions, it doesn't bifurcate the type: `Vec` is still `Vec` and
> so works with any crate expecting a `Vec`. Users of `Vec` can choose if they are in a path where allocation failures are recoverable (e.g., handling an
> incoming request), or non-recoverable (e.g., during application startup) without having to change the type they are using.
> 
> If these methods are not added to the standard library, then it is highly likely that the code bases that require fallible allocation will take one of these
> approaches themselves, forking the standard library if required.

Reviewer: Longer rational section emphasizing the key benefits of the current approach.

```suggestion
For the following reasons, we believe this the best option available at this time, and worthwhile to do rather than wait for a better option.
These are the reason why:

 - **Conforms to existing pattern**: Adding methods with the `try_` prefix follows the existing pattern within the standard library (as used by `Box::try_new` and `Vec::try_reserve`).
   They were added on a case-by-case basis for similar reasons. We are simply trying to push that effort further long without causing undue controversy.
 
 - **Works in Rust today**: It doesn't require any changes to the language or the compiler. Even if we knew exactly what change we would want (which we don't) coordinating the change and library modification afterwords is a long process.
   Until marquee use-cases like Rust in Linux are accepted upstream, we are hesitant to add uncertainty with high-latency projects.
  
 - **No new types splitting the ecosystem**: Although this design does bifurcate the functions, it doesn't bifurcate the type.
   `Vec` is still `Vec` and so works with any crate expecting a `Vec`.
   Multiple `Vec` types would hinder library interop when code one didn't write requires the other `Vec`.
   More functions / ad-hoc `try_` namespacing might be aesthetically displeasing, but doesn't levy any of these harsh forking penalties.

   Users of `Vec` can choose if they are in a path where allocation failures are recoverable (e.g., handling an incoming request), or non-recoverable (e.g., during application startup) without having to change the type they are using.

- **Usual beloved `Result` ergonomics**: Methods that requiring catching allocations failures more coarsely or out-of-band are more analogous to exception handling and errno, both of which are widely disliked by Rustaceans in comparison to `Result` and `?`.

- **Avoids forks, encourages interop** Conversely, if these methods are not added to the standard library, then it is highly likely that the code bases that require fallible allocation will take one of these approaches themselves, forking the standard library if required.
 Coalescing around a single fork across projects is unlikely, as it fostering a crates.io library ecosystem around such a fork.
 The longer we wait, the more low-level projects of this sort we are aiming for are likely to give up on reconcilling with the standard library, or sharing code with similar projects via libraries.
  
- **Roughly forwards compatible**: The ideal solution would avoid the `try_` ad-hoc namespacing without sacrificing these other benefits.
  While we don't know what that looks like on the *language* side, we are reasonably confident the *library* side will be quite similar.
  
  The essential insight of the current implementation is that we can share code between the old infallible and new fallible styles by using a polymorphic error type `Result<_, E>`.
  By making `E = !`, we "push" up the panicking as close to the original failure as possible, just like today, avoiding `Result`-propogating overhead when we're just going to end up panicking anyways.
  While we don't know the ideal form of this looks like *statically* in terms of types, coercions, and other compile-time features, we are pretty confident we do have what we want *dynamically* with the *behavior* of the old and new (`try_`) functions being what everyone wants.
  
  The current implementation provides both behaviors with minimal duplication, and so we suspect that whatever final form the interface takes, as long as it continues supporting these two desirable behaviors, our unified implementation an continue providing them.
```

Author: Thanks: I've merged this in with a few modifications.

in the long term: see the "Drawbacks" section below), instead we are suggesting this as a way forward to unblock important work (see the "Motivations" section
below) while we explore other alternatives. We have no plans to stabilize this feature in its current form and fully expect that these methods will be removed
in the future.

# Motivation
[motivation]: #motivation

The motivation for this change is well documented by other RFCs, such as the proposed [fallible_allocation feature](https://github.com/rust-lang/rfcs/pull/3140)
and the accepted [fallible_collection_alloc feature](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md).

As a brief summary, there are environments where dynamic allocation is required and failing to allocate is both expected and able to be handled. For example,
in OS Kernel environments (such as [Linux](https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/)), some embedded
systems, high-reliability systems (such as databases) and multi-user services (where a single request may fail without the entire service halting, such as a
web server).

> 

Reviewer 3: ```suggestion

Users in such environments may wish to make use of the `no_global_oom_handling` cfg added in <https://github.com/rust-lang/rust/pull/84266> in order to ban infallible allocation in their codebase's entire dependency graph. In such a situation they may still wish to use use the standard library's `Vec` type, but in combination with custom fallible allocation extension methods or a `FallibleVec` wrapper type provided by an external crate such as [fallible_collections](https://crates.io/crates/fallible_collections). However, once the `no_global_oom_handling` CFG is enabled, basic `Vec` methods like `push` are removed, making it difficult and unsafe to implement fallible equivalents outside of the standard library.

The additional methods proposed by this RFC would be enough for users in OOM-sensitive environments to make use of `no_global_oom_handling` without needing to fork or otherwise reimplement `Vec` externally. This is a sufficient stopgap solution for many use cases while a more complete fallible allocation solution is developed.

```

Per [my recent comment](https://github.com/rust-lang/rfcs/pull/3271#issuecomment-1279560444), and @ChrisDenton's suggestion to update the motivation section, here's my suggested update to the RFC text to capture the `no_global_oom_handling` use case. It looks like this was already mentioned in reference-level explanation, but hopefully adding it to the motivation section will also help clarify to readers how this RFC enables a use case that currently can't be accomplished.

Also sorry if the suggested wording isn't great, feel free to reword as needed if you agree this is worth adding!

Reviewer: Thanks! This looks good to me. @dpaoliello do you want to accept it?

This approach is referred to as "infallible allocation": from the perspective of a function caller the allocation can never fail, because if it does then the called function will `panic` and the caller won't live to observe the failure.

There are, however, environments where allocation failures are both expected and required to be handled more gracefully.
Consider a multi-user service such as a database or a web server: if a request comes in to the service that requires allocating more memory than is available on the system, then the service can respond with a failure for just that request, and it can continue to service other requests.
This approach is referred to as "fallible allocation": each function that may allocate needs to indicate to its caller if an attempt to allocate has failed, and that caller must explicitly handle that possibility.

Within the Rust standard library, one can identify "fallible allocation" functions by their names being prefixed with `try_` and their return types being an instance of `Result`.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> Currently any function in `alloc` that may all the global OOM handler (which panics) rather than let its caller handle all allocation failures is marked with `#[cfg(not(no_global_oom_handling))]` (see <https://github.com/rust-lang/rust/pull/84266>),

Reviewer 4: ```suggestion
Currently any function in `alloc` that may call the global OOM handler (which panics) rather than let its caller handle all allocation failures is marked with `#[cfg(not(no_global_oom_handling))]` (see <https://github.com/rust-lang/rust/pull/84266>),
```

- This requires `panic=unwind`, which might not be possible in the environment that requires fallible allocation (e.g., embedded systems or OS kernels).
- The standard library and 3rd party crates may not expect this pattern, and so have objects in invalid states at the point of OOM.

### Create a new "fallible allocation" `Vec` type

Instead of bifurcating the `Vec` methods, we could instead bifurcate the `Vec` type itself: thus we would keep `Vec` as the infallible version and introduce a
new fallible version (for the purposes of this RFC, let's call it `FallibleVec`). Under the covers, these two types could rely on the same implementation
type/functions to minimize code duplication.

Advantages:

- Avoids cluttering `Vec` with new APIs.
- Reduces decision points for developers: they only need to decide when choosing the type, rather than for each method call.

Disadvantages:

- Still requires bifurcating traits, and the functions that call those crates.
- `FallibleVec` cannot be used where a something explicitly asks for a `Vec` or any of the infallible allocation traits.

> - Requires a developer to choose at the type level if they need fallible or infallible allocation (unless there's a way to convert between the types).

Reviewer 5: I don't see how these are issues. `FallibleVec` could wrap `Vec` (or vice versa) and interconversion would be free. And borrows could be transmuted into each other for free too.

The types could exist at all times, even for code that configures infallible methods out of std. It can take a Vec in its public APIs but internally always turn it into a `&mut FallibleVec` and call those methods. This way its public interface can be compatible with both fallible and infallible code while providing less friction for the latter type.

> ### Always return `Result` (with the error based on the allocator), but allow implicit unwrapping for "infallible allocation"

Reviewer 5: Instead of being generic over the `Err` type of a `Result` wouldn't it be possible to make it generic over the return-value itself? This probably needs GATs to have something that produce the return type.

```rust
    pub fn try_push(&mut self, value: T) -> A::AllocationResult<()>::Output;
```

which then produces `()` or `Result<(), AllocError>` depending on the allocator.

> If the `Allocator` trait is updated to indicate what error is return if the allocation fails (with infallible allocation returning `!`), then the allocating

Reviewer 4: ```suggestion
If the `Allocator` trait is updated to indicate what error is returned if the allocation fails (with infallible allocation returning `!`), then the allocating
```
